### ISSUE TITLE
I hope the server-related code to be placed in a separate project.

### ISSUE THREAD

**Issue by tiandian** [chromelyapps/Chromely#123]
I found a lot of codes in project Chromely.CefGlue and Chromely.Core are about create web server and websocket server, which makes the Chromely project more difficult to understand.

In fact, I don't want to use these codes, because I will use asp.net core(maybe other webserver) to act as a webserver/websocket server. i will convert CefRequest to a  asp.net core request and send it to asp.net core pipeline and get the response then send back to cef.

So better make these codes(about web server and websocket server) in a separate project, make it replaceable, do this can make Chromely lighter and more readable.

thanks!

**Comment by mattkol** [comment]
@tiandian the Webserver like most features on Chromely are optional. You need to register it to use it. So, yes, already replaceable.

I am not sure putting it in a separate assembly will help. We still have to add it somehow ... right now it is in a separate folder - [ServerHandlers](https://github.com/chromelyapps/Chromely/tree/c8f9b2fb45b57670fd7322a9d1fbe67b39ec9782/src/Chromely.CefGlue/Browser/ServerHandlers).

It also looks like it is less used so, in my opinion it would either not be added for next releases or more isolated than it is now. So may be a separate project can help for that reason. Or we just let expert users use the code as they want.

Meanwhile, I will suggest you delete everything relating to the Webserver for your code base.

**Comment by tiandian** [comment]
Not only  the folder  ServerHandlers,  there are many other folder and files and methods are services for the Chromely's  bulit-in webserver,  when use other webserver, these codes are useless, e.g.:  

    Chromely.Core -> Infrastructure Folder->*.*
    Chromely.Core -> JsonMapper Folder->*.*
    Chromely.Core -> RestfulService Folder->*.*
    Chromely.Core -> MimeMapper.cs, IChromelyContainer.cs, IChromelyLogger.cs, IChromelyWebsocketHandler.cs

    Chromely.CefGlue->ServerHandlers->*.*
    Chromely.CefGlue->RestfulService->*.*
    Chromely.CefGlue->Handlers Folder->CefGlueHttpSchemeHandler.cs,CefGlueHttpSchemeHandlerFactory.cs,CefGlueMessageRouterHandler.cs,CefGlueResourceSchemeHandler.cs,CefGlueResourceSchemeHandlerFactory.cs,CefPostDataStream.cs
    Chromely.CefGlue->WebsocketMessageSender.cs, WebsocketServerRunner.cs

    And also some codes in other files are services for Chromely's bulit-in webserver, such as : window.RegisterServiceAssembly(); window.ScanAssemblies();


All above codes are  services for Chromely's bulit-in webserver,  are not ui related codes, If use other webserver,  they are useless, and it is difficult to remove any one of them, because there are not pluggable, remove anyone will lead compile error.

So i hope the codes for Chromely's bulit-in webserver(not ui related) can put in a separate project (e.g.: Chromely.WebServer) and make it pluggable, after move above code in a separate project, Chromely will be more lighter and readable.

thanks!

**Comment by mattkol** [comment]
@tiandian I see what you mean. We are not referring to the same thing. I was referring to WebsocketServer.

What you are referring to is at the Core of Chromely IPC - the Restful resource pattern approach. No, you cannot replace that - the way it is designed at the moment. And no, cannot be replaced with a any other server. They are used to translate messages between the Browser and the Renderer and vice-versa. It is a pattern, if you may, rather than a "service/server". In layman's term you can see the "Renderer" as the client and "Browser" as the server.

But yes, it could be in a different assembly. But note that it may also add unnecessary complexity. Some of those files you listed have dependencies on CefSharp or CefGlue, so to do it well and avoid cyclic dependency each must have it's own separate assembly for that purpose.

But we will see.

**Comment by tiandian** [comment]
`No, you cannot replace that - the way it is designed at the moment. And no, cannot be replaced with a any other server. `

No, it can be replaced with other server ,e.g., it can be replace with asp.net core, only thing to do is create  create a CefResourceHandler, overrite  ProcessRequest method, do as below:

        public class AspHttpSchemeHandler : CefResourceHandler
        {
            ...

            protected override bool ProcessRequest(CefRequest request, CefCallback callback)
            {
                 1. convert CefRequest to asp.net core HttpRequest
                 2. send HttpRequest to asp.net core pipline
                 3. get HttpResponse from asp.net core pipline
                 4. send HttpResponse back to Cef
            }
            ...
        }


I has a asp.net core program, asp.net core has more features than Chromely's bulit-in webserver, it is not easy to convert every asp.net core controller to Chromely's controller, 

So i hope Chromely should do more friendly to asp.net core(and other third-part webserver), and focus in linking to asp.net core , not focus in Chromely's bulit-in webserver,  Chromely's bulit-in webserver is still a little young for big project, hope users write code for Chromely's bulit-in webserver is not a good idea.

**Comment by mattkol** [comment]
@tiandian the way I see it we are not on the same page.

**Chromely does not care what you use for your source of data.**

Chromely is a framework. If you choose .NET ASP.NET Core or any other source that is your choice. You can choose to do that in the handler you described or in the Controller ... all your choice.  I think [this](https://github.com/chromelyapps/Chromely/issues/100) is an example of a developer trying to access an external server from the Controller. Please see my [comment -](https://github.com/chromelyapps/Chromely/issues/100#issuecomment-498642875)

Some of the files you listed earlier are are default handlers. This is ok for most developers. What you described above is a custom handler and you are allowed to do that. The wiki tells you how to register that ...

[Register handlers](https://github.com/chromelyapps/Chromely/wiki/Configuration#how-to-register-custom-scheme-handlers)

````csharp
var config = ChromelyConfiguration()
                              .Create()
                              ....
                              .RegisterSchemeHandler("http", "tiandian.com", new AspHttpSchemeHandler())
                              .....
````
When I say cannot be replaced at is at the moment, I am talking about the restful pattern. The functionalities are replaceable. They are in wikis and demos.

**Comment by mattkol** [comment]
@tiandian  just curious - any reason  you call the handlers built-in webserver? Never see anybody refer to them that way, but thinking about it some people will think they are.

In CEF world they are called handlers - scheme handlers specifically.
[CEF Wiki](https://bitbucket.org/chromiumembedded/cef/wiki/GeneralUsage#markdown-header-scheme-handler)
[Stackoverflow](https://stackoverflow.com/questions/35965912/cefsharp-custom-schemehandler)
[CefGlue page](https://bitbucket.org/xilium/xilium.cefglue/issues/73/http-s-custom-scheme-handler-crashes-cef)

**Comment by tiandian** [comment]
I know what you mean.

As for me, I only need a cross-platform cef-base ui framework, but in Chromely, most of the code are webserver-related code which i will not use,  so i think current Chromely is not for me.

So for me thebetter way is to use CefGlue directly, and maybe use part of Chromely's ui-relate code.

Thank you.
